### ISSUE TITLE
ConcurrentModificationException

### ISSUE THREAD

**Issue by 472574155** [awslabs/aws-mobile-appsync-sdk-android#53]
**Describe the bug**
When multiSubscription using, unexpectly exception happened.

**To Reproduce**
When net changed, easily happened

**Expected behavior**
java.util.ConcurrentModificationException
        at java.util.HashMap$HashIterator.nextEntry(HashMap.java:851)
        at java.util.HashMap$KeyIterator.next(HashMap.java:885)
        at com.amazonaws.mobileconnectors.appsync.subscription.RealSubscriptionManager$1.onError(RealSubscriptionManager.java:196)
        at com.amazonaws.mobileconnectors.appsync.subscription.mqtt.MqttSubscriptionClient$1.onFailure(MqttSubscriptionClient.java:85)
        at org.eclipse.paho.android.service.MqttTokenAndroid.notifyFailure(MqttTokenAndroid.java:146)
        at org.eclipse.paho.android.service.MqttAndroidClient.simpleAction(MqttAndroidClient.java:1501)
        at org.eclipse.paho.android.service.MqttAndroidClient.connectAction(MqttAndroidClient.java:1439)
        at org.eclipse.paho.android.service.MqttAndroidClient.onReceive(MqttAndroidClient.java:1368)
        at android.support.v4.content.LocalBroadcastManager.executePendingBroadcasts(LocalBroadcastManager.java:297)
        at android.support.v4.content.LocalBroadcastManager.access$000(LocalBroadcastManager.java:46)
        at android.support.v4.content.LocalBroadcastManager$1.handleMessage(LocalBroadcastManager.java:116)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6242)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)

**Environment(please complete the following information):**
 - AppSync SDK Version: [2.6.+]

**Device Information (please complete the following information):**
 - Device: SHAPE AQUOS SH-M05
 - Android Version: [Nougat 7.1.2]

**Comment by cbommas** [comment]
@472574155 
I double checked Maven and tried to reproduce the error, however I wasn't able to.
Can you share your build.gradle file?

My build.gradle contains 

          implementation "com.amazonaws:aws-android-sdk-appsync-gradle-plugin:2.6.+"
          implementation "com.amazonaws:aws-android-sdk-appsync:2.6.+"

and this pulls down the latest version when I do a gradle sync.

**Comment by 472574155** [comment]
@cbommas 
This issue still happened.
My build.gradle contains
`
implementation 'com.amazonaws:aws-android-sdk-appsync:2.6.26'
implementation 'com.amazonaws:aws-android-sdk-s3:2.7.2'
implementation 'com.amazonaws:aws-android-sdk-cognito:2.6.30'
implementation 'com.amazonaws:aws-android-sdk-mobile-client:2.6.31'
implementation 'com.amazonaws:aws-android-sdk-cognitoidentityprovider:2.7.2'
implementation 'com.amazonaws:aws-android-sdk-sns:2.6.+'

implementation 'com.amazonaws:aws-android-sdk-core:2.7.2'
implementation 'com.amazonaws:aws-android-sdk-ddb:2.6.31'
implementation 'com.google.firebase:firebase-messaging:17.3.4'
implementation 'com.google.firebase:firebase-core:16.0.4'

implementation 'org.eclipse.paho:org.eclipse.paho.client.mqttv3:1.2.0'
implementation 'org.eclipse.paho:org.eclipse.paho.android.service:1.1.1'
`

**Comment by 472574155** [comment]
@cbommas 
This issue still happened.
My build.gradle contains

implementation 'com.amazonaws:aws-android-sdk-appsync:2.6.26'
implementation 'com.amazonaws:aws-android-sdk-s3:2.7.2'
implementation 'com.amazonaws:aws-android-sdk-cognito:2.6.30'
implementation 'com.amazonaws:aws-android-sdk-mobile-client:2.6.31'
implementation 'com.amazonaws:aws-android-sdk-cognitoidentityprovider:2.7.2'
implementation 'com.amazonaws:aws-android-sdk-sns:2.6.+'

implementation 'com.amazonaws:aws-android-sdk-core:2.7.2'
implementation 'com.amazonaws:aws-android-sdk-ddb:2.6.31'
implementation 'com.google.firebase:firebase-messaging:17.3.4'
implementation 'com.google.firebase:firebase-core:16.0.4'

implementation 'org.eclipse.paho:org.eclipse.paho.client.mqttv3:1.2.0'
implementation 'org.eclipse.paho:org.eclipse.paho.android.service:1.1.1'

**Comment by 472574155** [comment]
@cbommas 
Have you checked the jar file? 
http://central.maven.org/maven2/com/amazonaws/aws-android-sdk-appsync/2.6.26/aws-android-sdk-appsync-2.6.26-sources.jar  (Is maven url right?)
1. download file aws-android-sdk-appsync-2.6.26-sources.jar
2. (.jar) -> (.zip)
3. unzip file
4.find file (aws-android-sdk-appsync-2.6.26-sources⁩ ▸ ⁨com⁩ ▸ ⁨amazonaws⁩ ▸ ⁨mobileconnectors⁩ ▸ ⁨appsync⁩ ▸ ⁨subscription⁩ ▸ RealSubscriptionManager.java)
5. Line No.196

**Comment by cbommas** [comment]
@472574155 
Thank you for pointing this out. I was able to track it down -the code in the github repo was updated after the v2.6.26 release as part of this PR https://github.com/awslabs/aws-mobile-appsync-sdk-android/pull/41 which contains the fix for the issue you are seeing.

Sorry about the confusion. I will get this released in the next few days and will post back on this thread once the release is completed.
